File: services/assessment-service/src/test/java/com/visor/school/assessment/service/AssessmentServiceTest.java
package com.visor.school.assessment.service;

import com.visor.school.assessment.model.Assessment;
import com.visor.school.assessment.model.AssessmentStatus;
import com.visor.school.assessment.model.AssessmentType;
import com.visor.school.assessment.repository.AssessmentRepository;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.math.BigDecimal;
import java.util.Arrays;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
class AssessmentServiceTest {

    @Mock
    private AssessmentRepository assessmentRepository;

    @InjectMocks
    private AssessmentService assessmentService;

    private UUID testClassId;
    private UUID testTeacherId;

    @BeforeEach
    void setup() {
        testClassId = UUID.randomUUID();
        testTeacherId = UUID.randomUUID();
        assessmentService = new AssessmentService(assessmentRepository);
    }

    @Test
    void shouldCreateAssessment() {
        // Given
        Assessment assessment = new Assessment(
            testClassId,
            "Midterm Exam",
            AssessmentType.EXAM,
            new BigDecimal("100.0"),
            testTeacherId
        );
        when(assessmentRepository.save(any(Assessment.class))).thenReturn(assessment);

        // When
        Assessment result = assessmentService.createAssessment(
            testClassId,
            "Midterm Exam",
            AssessmentType.EXAM,
            new BigDecimal("100.0"),
            testTeacherId,
            null,
            null,
            null
        );

        // Then
        assertNotNull(result);
        assertEquals("Midterm Exam", result.getName());
        assertEquals(AssessmentStatus.DRAFT, result.getStatus());
        verify(assessmentRepository).save(any(Assessment.class));
    }

    @Test
    void shouldGetAssessmentById() {
        // Given
        UUID assessmentId = UUID.randomUUID();
        Assessment assessment = new Assessment(
            testClassId,
            "Test",
            AssessmentType.TEST,
            new BigDecimal("100.0"),
            testTeacherId
        );
        when(assessmentRepository.findById(assessmentId)).thenReturn(Optional.of(assessment));

        // When
        Assessment result = assessmentService.getAssessment(assessmentId);

        // Then
        assertNotNull(result);
        verify(assessmentRepository).findById(assessmentId);
    }

    @Test
    void shouldReturnNullWhenAssessmentNotFound() {
        // Given
        UUID assessmentId = UUID.randomUUID();
        when(assessmentRepository.findById(assessmentId)).thenReturn(Optional.empty());

        // When
        Assessment result = assessmentService.getAssessment(assessmentId);

        // Then
        assertNull(result);
    }

    @Test
    void shouldPublishAssessment() {
        // Given
        UUID assessmentId = UUID.randomUUID();
        Assessment assessment = new Assessment(
            testClassId,
            "Test",
            AssessmentType.TEST,
            new BigDecimal("100.0"),
            testTeacherId
        );
        when(assessmentRepository.findById(assessmentId)).thenReturn(Optional.of(assessment));
        when(assessmentRepository.save(any(Assessment.class))).thenAnswer(invocation -> invocation.getArgument(0));

        // When
        Assessment result = assessmentService.publishAssessment(assessmentId);

        // Then
        assertEquals(AssessmentStatus.PUBLISHED, result.getStatus());
        verify(assessmentRepository).save(assessment);
    }

    @Test
    void shouldGetAssessmentsByClassId() {
        // Given
        Assessment assessment1 = new Assessment(
            testClassId,
            "Test 1",
            AssessmentType.TEST,
            new BigDecimal("100.0"),
            testTeacherId
        );
        Assessment assessment2 = new Assessment(
            testClassId,
            "Test 2",
            AssessmentType.QUIZ,
            new BigDecimal("50.0"),
            testTeacherId
        );
        when(assessmentRepository.findByClassId(testClassId)).thenReturn(Arrays.asList(assessment1, assessment2));

        // When
        List<Assessment> result = assessmentService.getAssessmentsByClass(testClassId);

        // Then
        assertEquals(2, result.size());
        verify(assessmentRepository).findByClassId(testClassId);
    }
}

File: services/academic-assessment-service/src/test/kotlin/com/visor/school/assessment/service/AssessmentServiceTest.kt
package com.visor.school.assessment.service

import com.visor.school.assessment.model.Assessment
import com.visor.school.assessment.model.AssessmentStatus
import com.visor.school.assessment.model.AssessmentType
import com.visor.school.assessment.repository.AssessmentRepository
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.junit.jupiter.MockitoExtension
import org.mockito.kotlin.*
import java.math.BigDecimal
import java.time.LocalDate
import java.util.Optional
import java.util.UUID

@ExtendWith(MockitoExtension::class)
class AssessmentServiceTest {

    @Mock
    private lateinit var assessmentRepository: AssessmentRepository

    @InjectMocks
    private lateinit var assessmentService: AssessmentService

    private val testClassId = UUID.randomUUID()
    private val testTeacherId = UUID.randomUUID()

    @BeforeEach
    fun setup() {
        assessmentService = AssessmentService(assessmentRepository)
    }

    @Test
    fun `should create assessment`() {
        // Given
        val assessment = Assessment(
            classId = testClassId,
            name = "Midterm Exam",
            type = AssessmentType.EXAM,
            totalPoints = BigDecimal("100.0"),
            createdBy = testTeacherId
        )
        whenever(assessmentRepository.save(any())).thenReturn(assessment)

        // When
        val result = assessmentService.createAssessment(
            classId = testClassId,
            name = "Midterm Exam",
            type = AssessmentType.EXAM,
            totalPoints = BigDecimal("100.0"),
            createdBy = testTeacherId
        )

        // Then
        assertNotNull(result)
        assertEquals("Midterm Exam", result.name)
        assertEquals(AssessmentStatus.DRAFT, result.status)
        verify(assessmentRepository).save(any())
    }

    @Test
    fun `should get assessment by id`() {
        // Given
        val assessmentId = UUID.randomUUID()
        val assessment = Assessment(
            id = assessmentId,
            classId = testClassId,
            name = "Test",
            type = AssessmentType.TEST,
            totalPoints = BigDecimal("100.0"),
            createdBy = testTeacherId
        )
        whenever(assessmentRepository.findById(assessmentId)).thenReturn(Optional.of(assessment))

        // When
        val result = assessmentService.getAssessment(assessmentId)

        // Then
        assertNotNull(result)
        assertEquals(assessmentId, result?.id)
        verify(assessmentRepository).findById(assessmentId)
    }

    @Test
    fun `should return null when assessment not found`() {
        // Given
        val assessmentId = UUID.randomUUID()
        whenever(assessmentRepository.findById(assessmentId)).thenReturn(Optional.empty())

        // When
        val result = assessmentService.getAssessment(assessmentId)

        // Then
        assertNull(result)
    }

    @Test
    fun `should publish assessment`() {
        // Given
        val assessmentId = UUID.randomUUID()
        val assessment = Assessment(
            id = assessmentId,
            classId = testClassId,
            name = "Test",
            type = AssessmentType.TEST,
            totalPoints = BigDecimal("100.0"),
            createdBy = testTeacherId
        )
        whenever(assessmentRepository.findById(assessmentId)).thenReturn(Optional.of(assessment))
        whenever(assessmentRepository.save(any())).thenAnswer { it.arguments[0] as Assessment }

        // When
        val result = assessmentService.publishAssessment(assessmentId)

        // Then
        assertEquals(AssessmentStatus.PUBLISHED, result.status)
        verify(assessmentRepository).save(assessment)
    }

    @Test
    fun `should get assessments by class id`() {
        // Given
        val assessment1 = Assessment(
            classId = testClassId,
            name = "Test 1",
            type = AssessmentType.TEST,
            totalPoints = BigDecimal("100.0"),
            createdBy = testTeacherId
        )
        val assessment2 = Assessment(
            classId = testClassId,
            name = "Test 2",
            type = AssessmentType.QUIZ,
            totalPoints = BigDecimal("50.0"),
            createdBy = testTeacherId
        )
        whenever(assessmentRepository.findByClassId(testClassId)).thenReturn(listOf(assessment1, assessment2))

        // When
        val result = assessmentService.getAssessmentsByClass(testClassId)

        // Then
        assertEquals(2, result.size)
        verify(assessmentRepository).findByClassId(testClassId)
    }
}

